driver/src/price_estimation/kraken/api.rs

```diff
@@ -0,0 +1,272 @@
+    }
+
+    fn asset_pairs(&self) -> Result<HashMap<String, AssetPair>> {
+        self.client
+            .get(format!("{}/AssetPairs", self.base_url))
+            .context("failed to retrieve list of asset pairs from Kraken")?
+            .json::<KrakenResult<_>>()
+            .context("failed to parse asset pairs JSON")?
+            .into_result()
+    }
+
+    fn ticker(&self, pairs: &[&str]) -> Result<HashMap<String, TickerInfo>> {
+        self.client
+            .get(format!("{}/Ticker?pair={}", self.base_url, pairs.join(",")))
+            .context("failed to retrieve ticker infos from Kraken")?
+            .json::<KrakenResult<_>>()
+            .context("failed to parse ticker JSON")?
+            .into_result()
+    }
+}
+
+/// The result type that is returned by Kraken on API requests. This type is
+/// only used internally.
+#[derive(Clone, Debug, Deserialize)]
+struct KrakenResult<T> {
+    error: Vec<String>,
+    result: Option<T>,
+}
+
+impl<T> KrakenResult<T> {
+    fn into_result(self) -> Result<T> {
+        if let Some(result) = self.result {
+            Ok(result)
+        } else if !self.error.is_empty() {
+            Err(anyhow!("Kraken API errors: {:?}", self.error))
+        } else {
+            Err(anyhow!("unknown Kraken API error"))
+        }
+    }
+}
+
+/// A struct representing an asset retrieved from the Kraken API.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct Asset {
+    pub altname: String,
+}
+
+impl Asset {
+    /// Create a new asset from an alternate name.
+    #[cfg(test)]
+    pub fn new(altname: &str) -> Asset {
+        Asset {
+            altname: altname.into(),
+        }
+    }
+}
+
+/// A struct representing an asset pair retrieved from the Kraken API.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct AssetPair {
+    pub base: String,
+    pub quote: String,
+}
+
+impl AssetPair {
+    /// Create a new asset pair from base and quote assets.
+    #[cfg(test)]
+    pub fn new(base: &str, quote: &str) -> AssetPair {
+        AssetPair {
+            base: base.into(),
+            quote: quote.into(),
+        }
+    }
+}
+
+/// A struct representing ticker info for an asset pair including price
+/// information.
+///
+/// Note that this is only a small subset of the data provided by the Kraken API
+/// and only the parts required for retrieving price estimates for the solver
+/// are included.
+#[derive(Clone, Debug, Deserialize)]
+#[cfg_attr(test, derive(PartialEq))]
+pub struct TickerInfo {
+    pub p: PricePair,
+}
```

Comment:
I am confused by the name of this struct, since nothing about it seems to have anything to do with ticking of any sort.

Reply:
I think the word "ticker" is related to stock tickers, and represents the price widgets you see in the Kraken interface. The name was chosen based on the Kraken API documentation:
https://www.kraken.com/features/api#get-ticker-info